### PR TITLE
hal: wifi: add support of wifi subsystem for esp32c3

### PR DESCRIPTION
**Pull request by uLipe**
By adding wifi adapter and wpa_supplicant files.

Signed-off-by: Felipe Neves <felipe.neves@espressif.com>

### CODE REVIEW

**Comment by uLipe** [comment]
@sylvioalves @glaubermaroto PTAL

File: zephyr/esp32c3/src/bt/esp_bt_adapter.c [hunk 1]
```diff
@@ -31,7 +31,7 @@
 #include <zephyr.h>
 #include <sys/printk.h>
 #include <random/rand32.h>
-#include <drivers/interrupt_controller/intc_esp32c3.h>
+//#include <drivers/interrupt_controller/intc_esp32c3.h>
```
Comment (glaubermaroto) [inline zephyr/esp32c3/src/bt/esp_bt_adapter.c:34]: please, remove the commented line.

File: components/esp_timer/src/esp_timer.c [hunk 2]
```diff
@@ -43,12 +43,16 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
...
+#include "esp32/rom/ets_sys.h"
 #include "esp32/rtc.h"
 #elif CONFIG_IDF_TARGET_ESP32S2
+#include "esp32s2/rom/ets_sys.h"
 #include "esp32s2/rtc.h"
 #elif CONFIG_IDF_TARGET_ESP32S3
+#include "esp32s3/rom/ets_sys.h"
 #include "esp32s3/rtc.h"
 #elif CONFIG_IDF_TARGET_ESP32C3
+#include "esp32c3/rom/ets_sys.h"
```
Comment (glaubermaroto) [inline components/esp_timer/src/esp_timer.c:55]: Is it necessary to include the same header to the other SoCs (aside esp32c3)? If so, why?

File: zephyr/esp32c3/src/common/esp_read_mac.c [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ * Copyright (c) 2021 Espressif Systems (Shanghai) Co., Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "esp_system.h"
+#include <rom/efuse.h>
+#include <zephyr.h>
+
+esp_err_t esp_read_mac(uint8_t *mac, esp_mac_type_t type)
```
Comment (glaubermaroto) [inline zephyr/esp32c3/src/common/esp_read_mac.c:11]: For both esp32 and esp32s2, this functon is currently defined under `zephyr/esp_shared/src/common/esp_system_api.c`. Should we make the same for esp32c3?

Comment (uLipe) [reply zephyr/esp32c3/src/common/esp_read_mac.c:11]: It is a bit different in the C3, I have plans of make this function common for it as well but in a later MR :)

File: components/esp_timer/src/esp_timer.c (hunk 2 above)
Comment (uLipe) [reply components/esp_timer/src/esp_timer.c:55]: For C3 is necessary, for the others I just placed them just to keep the consistency.

File: zephyr/esp32c3/CMakeLists.txt [hunk 4]
```diff
@@ -60,6 +67,10 @@ if(CONFIG_SOC_ESP32C3)
 
   zephyr_sources(
     ../../components/soc/esp32c3/gpio_periph.c
+    ../../components/esp_timer/src/ets_timer_legacy.c
+    ../../components/esp32c3/clk.c
```
Comment (sylvioalves) [inline zephyr/esp32c3/CMakeLists.txt:71]: This file has been updated. Can you rebase with latest version and test if it still works?

File: components/esp_timer/src/esp_timer.c (hunk 2 above)
Comment (glaubermaroto) [reply components/esp_timer/src/esp_timer.c:55]: If it is not necessary for the other SoCs, I suggest you consider not including it, otherwise, it may clutter the PR's history.

When another SoC needs this header, it can include in its own context. Does that make sense?

File: zephyr/esp32c3/src/common/esp_read_mac.c (hunk 3 above)
Comment (glaubermaroto) [reply zephyr/esp32c3/src/common/esp_read_mac.c:11]: Alternatively, you could consider unifying this function call in this same PR but in another commit, as done here:

https://github.com/zephyrproject-rtos/hal_espressif/pull/60/commits/9fcbf539faeee1220f57f56c82eda893f12912c6#diff-d3ff68eb35224ac623662cbaf1b15b2ca3d8131bc799c49b8c1550069ba79d3c

File: components/esp_timer/src/esp_timer.c (hunk 2 above)
Comment (sylvioalves) [reply components/esp_timer/src/esp_timer.c:55]: Let me give my 5 cents in here: `esp32XX/rom/ets_sys.h` is not necessary at all. I think you can remove all entries in `esp_timer.c` regarding `ets_sys.h`. Can you re-check this?

File: zephyr/esp32c3/src/common/esp_read_mac.c (hunk 3 above)
Comment (uLipe) [reply zephyr/esp32c3/src/common/esp_read_mac.c:11]: I considered but it will take more time than just drop esp32c3 definitions, even the EFUSE addresses are not properly documented. I really prefer to do this in a separate PR later.

File: zephyr/esp32c3/src/bt/esp_bt_adapter.c [hunk 5]
```diff
@@ -31,7 +31,6 @@
 #include <zephyr.h>
 #include <sys/printk.h>
 #include <random/rand32.h>
-#include <drivers/interrupt_controller/intc_esp32c3.h>
```
Comment (sylvioalves) [inline zephyr/esp32c3/src/bt/esp_bt_adapter.c:34]: OK, you removed it from here so that BT uses direct interrupt calls (as in WiFi), right?  
This could be a separate commit in this MR (touching BT stuff).

**Comment by sylvioalves** [comment]
LGTM

File: zephyr/esp32c3/src/bt/esp_bt_adapter.c (hunk 5 above)
Comment (uLipe) [reply zephyr/esp32c3/src/bt/esp_bt_adapter.c:34]: Yes, it will be added back when interrupt allocator gets merged.

**Review by sylvioalves — APPROVED** [review]
